readium/adapters/pdfium/pdfium-navigator/src/main/java/org/readium/adapters/pdfium/navigator/PdfiumEngineProvider.kt

```diff
@@ -22,19 +24,22 @@ import org.readium.r2.shared.publication.Publication
  */
 @ExperimentalReadiumApi
 public class PdfiumEngineProvider(
-    private val listener: PdfiumDocumentFragment.Listener? = null,
     private val defaults: PdfiumDefaults = PdfiumDefaults()
-) : PdfEngineProvider<PdfiumSettings, PdfiumPreferences, PdfiumPreferencesEditor> {
-
-    override suspend fun createDocumentFragment(input: PdfDocumentFragmentInput<PdfiumSettings>): PdfiumDocumentFragment =
-        PdfiumDocumentFragment(
-            publication = input.publication,
-            link = input.link,
-            initialPageIndex = input.initialPageIndex,
-            settings = input.settings,
-            appListener = listener,
-            navigatorListener = input.listener
-        )
+) : PdfEngineProvider<PdfiumDocumentFragment, PdfiumDocumentFragment.Listener, PdfiumSettings, PdfiumPreferences, PdfiumPreferencesEditor> {
+
+    override fun createDocumentFragmentFactory(
+        input: PdfDocumentFragmentInput<PdfiumDocumentFragment.Listener, PdfiumSettings>
+    ): SingleFragmentFactory<PdfiumDocumentFragment> =
```

Comment:
We return a `SingleFragmentFactory` which is a custom implementation of `FragmentFactory` able to instantiate a new instance without knowing the class of the fragment.